llvm/lib/CodeGen/CodeGenPrepare.cpp

```diff
@@ -2686,8 +2686,9 @@ bool CodeGenPrepare::dupRetToEnableTailCallOpts(BasicBlock *BB,
             attributesPermitTailCall(F, CI, RetI, *TLI)) {
           // Either we return void or the return value must be the first
           // argument of a known intrinsic or library function.
-          if (!V || (isIntrinsicOrLFToBeTailCalled(TLInfo, CI) &&
-                     V == CI->getArgOperand(0))) {
+          if (!V || isa<UndefValue>(V) ||
```

Comment:
Could we just leave V as nullptr back at line 2602?

Reply:
I think the explicit UndefValue check is fine. Especially if we also want to extend the phi case above to handle undef incoming values later.

Reply:
Shall we ever hit the phi case? Looking at https://clang.godbolt.org/z/eq8vTxvzM, it seems that in the first case it suffices one non-undef incoming value to tail call, whereas the second case is refined by CGP as follows:
```llvm
define dso_local i32 @foo() local_unnamed_addr #0 {
  %1 = load i32, ptr @i, align 4
  switch i32 %1, label %6 [
    i32 2, label %2
    i32 5, label %4
  ]

2:                                                ; preds = %0
  %3 = tail call i32 @bar() #2
  br label %6

4:                                                ; preds = %0
  %5 = tail call i32 @qux() #2
  ret i32 %5

6:                                                ; preds = %2, %0
  ret i32 undef
}
```
Thus getting tail called as part of this change.